File: packages/typia/src/schemas/metadata/MetadataConstantValue.ts
import { ClassProperties } from "../../typings/ClassProperties";

import { IJsDocTagInfo } from "./IJsDocTagInfo";
import { IMetadataConstantValue } from "./IMetadataConstantValue";
import { IMetadataTypeTag } from "./IMetadataTypeTag";

export class MetadataConstantValue {
  public readonly value: boolean | bigint | number | string;
  public tags: IMetadataTypeTag[][];
  public readonly description?: string | null;
  public readonly jsDocTags?: IJsDocTagInfo[];
  private name_?: string;

  private constructor(props: ClassProperties<MetadataConstantValue>) {
    this.value = props.value;
    this.tags = props.tags;
    this.description = props.description;
    this.jsDocTags = props.jsDocTags;
  }

  public static create(
    props: ClassProperties<MetadataConstantValue>,
  ): MetadataConstantValue {
    return new MetadataConstantValue(props);
  }

  public static from(json: IMetadataConstantValue<any>): MetadataConstantValue {
    return MetadataConstantValue.create({
      value: typeof json.value === "bigint" ? BigInt(json.value) : json.value,
      tags: json.tags,
      description: json.description,
      jsDocTags: json.jsDocTags,
    });
  }

  public getName(): string {
    return (this.name_ ??= getName(this));
  }

  public toJSON(): IMetadataConstantValue<any> {
    return {
      value:
        typeof this.value === "bigint" ? this.value.toString() : this.value,
      tags: this.tags,
      description: this.description,
      jsDocTags: this.jsDocTags,
    };
  }
}

const getName = (obj: MetadataConstantValue): string => {
  const base: string =
    typeof obj.value === "string"
      ? JSON.stringify(obj.value)
      : obj.value.toString();
  if (!obj.tags?.length) return base;
  const rows: string[] = obj.tags.map((row) => {
    const str: string = row.map((t) => t.name).join(" & ");
    return row.length === 1 ? str : `(${str})`;
  });
  return `(${base} & (${rows.join(" | ")}))`;
};

File: packages/typia/native/core/schemas/metadata/MetadataConstantValue.go
package metadata

import (
  "encoding/json"
  "fmt"
)

type IJsDocTagInfo struct {
  Name string
  Text []IJsDocTagInfo_IText
}

type IJsDocTagInfo_IText struct {
  Text string
  Kind string
}

type IMetadataSchema_IConstant_IValue struct {
  Value       any
  Tags        [][]IMetadataTypeTag
  Description *string
  JsDocTags   []IJsDocTagInfo
}

type MetadataConstantValue struct {
  Value       any
  Tags        [][]IMetadataTypeTag
  Description *string
  JsDocTags   []IJsDocTagInfo
  name_       string
}

func MetadataConstantValue_create(props MetadataConstantValue) *MetadataConstantValue {
  return &MetadataConstantValue{
    Value:       props.Value,
    Tags:        props.Tags,
    Description: props.Description,
    JsDocTags:   props.JsDocTags,
  }
}

func MetadataConstantValue_from(json IMetadataSchema_IConstant_IValue) *MetadataConstantValue {
  return MetadataConstantValue_create(MetadataConstantValue{
    Value:       json.Value,
    Tags:        json.Tags,
    Description: json.Description,
    JsDocTags:   json.JsDocTags,
  })
}

func (obj *MetadataConstantValue) GetName() string {
  if obj.name_ == "" {
    obj.name_ = metadataConstantValue_getName(obj)
  }
  return obj.name_
}

func (obj *MetadataConstantValue) ToJSON() IMetadataSchema_IConstant_IValue {
  return IMetadataSchema_IConstant_IValue{
    Value:       obj.Value,
    Tags:        obj.Tags,
    Description: obj.Description,
    JsDocTags:   obj.JsDocTags,
  }
}

func metadataConstantValue_getName(obj *MetadataConstantValue) string {
  return taggedName(metadataConstantValue_base(obj.Value), obj.Tags)
}

func metadataConstantValue_base(value any) string {
  if str, ok := value.(string); ok {
    data, err := json.Marshal(str)
    if err == nil {
      return string(data)
    }
  }
  return fmt.Sprint(value)
}
